Add arguments with effects

starkillerOG opened the pull request:
## Description:

Hyperion: Add posibility for arguments with effects.
If the turn on service is called with the effect parameter, now effect arguments of hyperion can be used in the call following this syntax:
- service: light.turn_on
   data:
       entity_id: "light.tv_light"
       effect: "EFFECT_NAME||ARGUMENT1=VALUE1||ARGUMENT2=VALUE2"

so first the name of the effect, then a seperation sign "||" then the name of the first argument followed by the value of this argument seperated with an "=" sign. A unlimeted amount of argumets can follow each with a "||" between arguments and a "=" between the argument and its value.

For instance two working effects for Hyperion including arguments would be:
effect: "Cinema brighten lights||color-start=[ 0, 0, 0 ]||color-end=[ 255, 136, 0 ]||fade-time=60.0"
or 
effect: "Knight rider||speed=1.0||fadeFactor=0.7||color=[0,255,0]"


**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation :** home-assistant/home-assistant.github.io#5037

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

starkillerOG committed:
Hyperion: Add arguments with effects

Hyperion: Add posibility for arguments with effects.
If the turn on service is called with the effect parameter, now effect arguments of hyperion can be used in the call following this syntax:
- service: light.turn_on
   data:
       entity_id: "light.tv_light"
       effect: "EFFECT_NAME||ARGUMENT1=VALUE1||ARGUMENT2=VALUE2"

so first the name of the effect, then a seperation sign "||" then the name of the first argument followed by the value of this argument seperated with an "=" sign. A unlimeted amount of argumets can follow each with a "||" between arguments and a "=" between the argument and its value.

For instance two working effects for Hyperion including arguments would be:
effect: "Cinema brighten lights||color-start=[ 0, 0, 0 ]||color-end=[ 255, 136, 0 ]||fade-time=60.0"
or 
effect: "Knight rider||speed=1.0||fadeFactor=0.7||color=[0,255,0]"

starkillerOG committed:
white spaces

starkillerOG committed:
Replace eval with ast.literal_eval

starkillerOG commented:
@MartinHjelmare @balloob could one of you review and merge this PR, its just a few lines?

balloob commented:
😠 Don't tag people for comments, there are 121 open PRs.

We should never use `literal_eval`.

starkillerOG commented:
@balloob do you have any sugestions what I can use instead of `ast.literal_eval`?

The values of the arguments are in string type, and can be numbers, tuples, booleans or lists. It depends on what effect you call and what arguments you want to supply.
Somehow I need to get rid of the string indicators `' ' ` around the values of the arguments inside the json blob supplied to Hyperion.

The only things that `ast.literal_eval` will accept are `strings, bytes, numbers, tuples, lists, dicts, sets, booleans, and None`. So that seemed reasonable for me, but I am happy to use something else, but I just do not know what.

starkillerOG committed:
change ast.literal_eval() to json.loads()

starkillerOG commented:
@balloob sorry, I found a solution, now using `json.loads()` instead of `ast.literal_eval()`

thundergreen commented:
I hope there will be a good documentation about arguments

balloob commented:
Don't use json load either. Service data can already contain any data structure.

starkillerOG commented:
@thundergreen their is already documentation about the effect arguments on the hyperion site, I added these two lines to the documentation of Home Assistant to refer to that documentation:

More about effect arguments can be found on [hyperion-project.org](https://hyperion-project.org/wiki/hyperion-remote-JSON-interface).
All specific effect arguments for each effect can be found in the json files on the [Hyperion Github](https://github.com/hyperion-project/hyperion/tree/master/effects).

starkillerOG commented:
@pvizeli and @balloob I do agree that defining a new service that has multiple arguments were you can just input the effect arguments in a diffrent parameter would be ideal. But I have now idea how I could do that and how that works. Therefore I used the already build in effect parameter that can be supplied to the build in light.turn_on service.

The drawback is that I have to use one parameter to supply both the argument name and all effect arguments with their corresponding values. So that is multiple types of data that i have to supplie (strings, numbers bolean etc in one parameter).

Is it possible to input a dictionarry in the build in light.turn_on effect parameter?
Than I could use a dictionarry to supply everything instead of a string with seperators ( || and = ).

Although I think that a dictonary would be a more complex syntax than the string with seperators that I have now.

What do you think that the best approach would be @balloob?

starkillerOG committed:
switch to dict instead of string

Using a dict instead of a string as data input

starkillerOG committed:
going back to json.loads

Tryed using a dict but it did not work because effect forces a string so the dict wil be read as string

starkillerOG commented:
@balloob I tyred supplying a dict to the effect argument like this service call:
```
{
  "entity_id": "light.tv_light",
  "effect": {"name": "Knight rider", "color" : [0,255,0]}
}
```
However this gave me a string instead of the expected dict, so still having the same problem.
I tryed defining a custom service, but I do not how to do that, could you help, or provide a hint?

balloob commented:
I guess you're right, it's currently not possible to do any effect.

I suggest that we close this PR for now. Eventually we're going to remove effect from `turn_on` / `turn_off` and make it a standalone service, but that requires a bit more work.

starkillerOG commented:
@balloob can't we merge this PR as is with the json.loads, such that effect arguments will work for now?
And impove it in the future with a standalone service as you indicated.

balloob commented:
No. We never allow services to receive data that has to be parsed as JSON, YAML or any other value.

starkillerOG commented:
Alright, thanks for your quick replay!
If you have time @balloob, could you take a look at my PR for the denonavr media player compent:
https://github.com/home-assistant/home-assistant/pull/13706
Submitted it 11 days ago after the denonavr library got updated with my new code supporting sound modes, but did not get any feedback yet on the HomeAssistant PR.
